### PR TITLE
PR: Deque RAII Implementation and Performance Optimization

### LINKED ISSUE #52: [Deque] RAII via `buffer_guard` not fully implemented due to missing `set_constructed_count` updates
**Problem:**
In the current implementation of `deque`'s memory management, `buffer_guard` is used to provide RAII for node allocations. However, `set_constructed_count` is not being actively updated in many places, which means that the RAII mechanism is **not fully effective**.

This is particularly problematic when using internal memory helper functions, for example:

```cpp
for (size_type i = 0; i < num_nodes; ++i) {
    *(start_node + i) = bufs_guard[i].get();
}

iterator new_start(start_node, *start_node + ((num_nodes * _buffer_size()) - n) / 2);

auto new_finish = std::uninitialized_fill_n(new_start, n, value);
```

In these cases, `set_constructed_count` is not updated after placement, so `buffer_guard` cannot correctly track which objects have been constructed. This may lead to **memory leaks or improper destruction** if an exception occurs, and effectively means that RAII is not functioning as intended.

### PR DESCRIPTION
**Pull request by jaehyung409**
## Summary
RAII-based memory management and performance optimization for Deque container with enhanced exception safety.
- Fixes #52 

### Features
- **RAII Pattern**: Comprehensive buffer_guard implementation for automatic resource management
- **Helper Functions**: Abstraction layer for range insertion operations
- **Memory Optimization**: Random-access iterator specialization with raw pointer compatibility

### Fixes
- **Reallocation Logic**: Fixed reallocate_map using correct node count
- **Insertion Position**: Corrected insertion positions in shift operations
- **Distance Calculation**: Enhanced compatibility with static_cast
- **Clear Operation**: Fixed nullptr handling in start_node management

### Testing & Benchmarking
- **Test Coverage**: Deque 81% (branch 62%)
- **Comprehensive Tests**: Basic, edge case, large dataset, and special operation tests
- **Stress Testing**: Large-scale operations and boundary condition validation

## Technical Details
- **Exception Safety**: RAII-based buffer management throughout all operations
- **Performance**: Optimized memory manipulation for random-access iterators
- **Code Quality**: Enhanced modularity with helper function abstractions

### CODE REVIEW

File: modules/datastructures/Deque/deque.cppm [hunk 1]
```diff
@@ -720,86 +747,107 @@ deque<T, Allocator>::size_type deque<T, Allocator>::calc_move_backward_now(Input
...
-                uninitialized_move_n_contiguous(alloc, first._current, move_now, dest._current);
-            } else {
-                uninitialized_move_n_contiguous(alloc, first, move_now, dest._current);
+    if constexpr (std::is_same_v<InputIter, iterator> || std::is_same_v<InputIter, const_iterator>) {
+        if constexpr (std::is_trivially_copyable_v<T>) {
+            while (count > 0) {
+                const size_type source_buffer_remaining = _buffer_size() - (first._current - first._first);
+                const size_type move_now = std::min(count, source_buffer_remaining);
+                std::memmove(dest, std::to_address(first), move_now * sizeof(T));
+                guard.add_constructed_count(move_now);
```
[the remark this answers is not in the document]
Comment (jaehyung409) [reply modules/datastructures/Deque/deque.cppm:799]: @copilot I already mentioned, caller must check the bound of buffer

File: modules/datastructures/Deque/deque.cppm [hunk 2]
```diff
@@ -1739,7 +2173,11 @@ template <class T, class Allocator> void deque<T, Allocator>::clear() noexcept {
             std::allocator_traits<buf_allocator>::destroy(_buf_alloc, std::to_address(it));
         }
     }
+
     buf *start_node = _map + _map_capacity / 2;
+    if (*start_node == nullptr) {
+        std::swap(*start_node, *_start._node);
```
[the remark this answers is not in the document]
Comment (jaehyung409) [reply modules/datastructures/Deque/deque.cppm:2179]: Allocate buffer is expensive operation. So we need to reuse node.

File: modules/datastructures/Deque/deque.cppm [hunk 3]
```diff
@@ -720,86 +747,107 @@ deque<T, Allocator>::size_type deque<T, Allocator>::calc_move_backward_now(Input
...
-        while (count > 0) {
-            const size_type buffer_remaining = _buffer_size() - (first._current - first._first);
-
-            const size_type fill_now = std::min(count, buffer_remaining);
-
-            uninitialized_fill_n_contiguous(alloc, first._current, fill_now, value);
-
-            first += fill_now;
-            count -= fill_now;
+    if constexpr (std::is_trivially_constructible_v<T> && std::is_trivially_copyable_v<T> &&
```
Comment (jaehyung409) [inline modules/datastructures/Deque/deque.cppm]:
```suggestion
    if constexpr (std::is_integral_v<T> && std::is_trivially_copyable_v<T> &&
```

File: modules/datastructures/Deque/deque.cppm [hunk 4]
```diff
@@ -209,47 +225,57 @@ export template <class T, class Allocator = std::allocator<T>> class deque {
     }
 
     template <class... Args>
-    void _shift_left_and_emplace(const difference_type distance_from_begin, iterator emplace_pos, Args &&...args);
+    void _shift_left_and_emplace(const size_type distance_from_begin, iterator emplace_pos, Args &&...args);
```
Comment (jaehyung409) [inline modules/datastructures/Deque/deque.cppm]:
```suggestion
    void _shift_left_and_emplace(size_type distance_from_begin, iterator emplace_pos, Args &&...args);
```

File: modules/datastructures/Deque/deque.cppm [hunk 5]
```diff
@@ -209,47 +225,57 @@ export template <class T, class Allocator = std::allocator<T>> class deque {
...
 
     template <class... Args>
-    void _shift_right_and_emplace(const difference_type distance_from_end, iterator emplace_pos, Args &&...args);
+    void _shift_right_and_emplace(const size_type distance_from_end, iterator emplace_pos, Args &&...args);
```
Comment (jaehyung409) [inline modules/datastructures/Deque/deque.cppm]:
```suggestion
    void _shift_right_and_emplace(size_type distance_from_end, iterator emplace_pos, Args &&...args);
```

File: modules/datastructures/Deque/deque.cppm [hunk 6]
```diff
@@ -209,47 +225,57 @@ export template <class T, class Allocator = std::allocator<T>> class deque {
...
 
-    void _shift_left_and_insert(const T &value, const difference_type distance_from_begin, iterator emplace_pos);
+    void _shift_left_and_insert(const T &value, const size_type distance_from_begin, iterator insert_pos);
```
Comment (jaehyung409) [inline modules/datastructures/Deque/deque.cppm]:
```suggestion
    void _shift_left_and_insert(const T &value, size_type distance_from_begin, iterator insert_pos);
```

File: modules/datastructures/Deque/deque.cppm [hunk 7]
```diff
@@ -209,47 +225,57 @@ export template <class T, class Allocator = std::allocator<T>> class deque {
...
 
-    void _shift_right_and_insert(const T &value, const difference_type distance_from_end, iterator emplace_pos);
+    void _shift_right_and_insert(const T &value, const size_type distance_from_end, iterator insert_pos);
```
Comment (jaehyung409) [inline modules/datastructures/Deque/deque.cppm]:
```suggestion
    void _shift_right_and_insert(const T &value, size_type distance_from_end, iterator insert_pos);
```
